src/dftbp/dftbplus/parser.F90

```diff
@@ -8488,13 +8362,139 @@ subroutine parseChimes(root, chimesRepInput)
+            inp%coupling = deltaT / inp%coupling
+          else
+            call error("Either CouplingStrength or Timescale must be set for Berendsen thermostats.")
+          end if
+        end if
+      end associate
+
+    case ("nosehoover")
+
+      thermostatInp%thermostatType = thermostatTypes%nhc
+      allocate(thermostatInp%nhc)
+      associate (inp => thermostatInp%nhc)
+        call readTempOrTempProfile_(thermNode, maxRun, tempProfileInp)
+        call getChildValue(thermNode, "CouplingStrength", inp%coupling, modifier=modifier,&
+            & child=child2)
+        call convertUnitHsd(char(modifier), freqUnits, child2, inp%coupling)
+
+        call getChildValue(thermNode, "ChainLength", inp%chainLength, 3)
+        call getChildValue(thermNode, "Order", inp%expOrder, 3)
+        call getChildValue(thermNode, "IntegratorSteps", inp%nExpSteps, 1)
+        call getChild(thermNode, "Restart",  child=child2, requested=.false.)
+        if (associated(child2)) then
+          allocate(inp%xnose(inp%chainLength))
+          allocate(inp%vnose(inp%chainLength))
+          allocate(inp%gnose(inp%chainLength))
+          call getChildValue(child2,"x",inp%xnose)
+          call getChildValue(child2,"v",inp%vnose)
+          call getChildValue(child2,"g",inp%gnose)
+        end if
+      end associate
+
+    case ("andersen")
+
+      thermostatInp%thermostatType = thermostatTypes%andersen
+      allocate(thermostatInp%andersen)
+      associate (inp => thermostatInp%andersen)
+        call readTempOrTempProfile_(thermNode, maxRun, tempProfileInp)
+        call getChildValue(thermNode, "ReselectProbability", inp%rescaleProb, child=child2)
+        if (inp%rescaleProb <= 0.0_dp .or. inp%rescaleProb > 1.0_dp) then
+          call detailedError(child2, "ReselectProbability must be in the range (0,1]!")
+        end if
+        call getChildValue(thermNode, "ReselectIndividually", inp%rescaleIndiv)
+      end associate
+
+    case ("none")
+
+      ! Create a fake thermostat with a single constant temperature value
+      ! It will only used to generate the initial velocities for the MD anyway.
+      thermostatInp%thermostatType = thermostatTypes%dummy
+      tempProfileInp%tempInts = [huge(1)]
+      tempProfileInp%tempMethods = [tempProfileTypes%constant]
+      tempProfileInp%tempValues = [minTemp]
+      if (.not. hasInitVelocities) then
+        ! Initial velocities had not been provided, overwrite 'safe' default value (minTemp)
+        ! by reading the temperature explicitly (needed for generating the initial velocities)
+        call readMDInitTemp(thermNode, tempProfileInp%tempValues(1), minTemp)
+      end if
+
+    case default
+      call getNodeHSDName(thermNode, thermName)
+      call detailedError(child, "Invalid thermostat '" // char(thermName) // "'")
+
+    end select
+
+  contains
+
+    subroutine readTempOrTempProfile_(thermNode, maxRun, tempProfileInp)
```

Comment:
comments...

Reply:
Sure, but being an internal subroutine of a subroutine, I'd suggest not to document every argument separately. I'll add a comment for the subroutine declaration, though.